README.md
# nettime

[![npm](https://img.shields.io/npm/v/nettime)](https://www.npmjs.com/package/nettime#top)
[![codecov](https://codecov.io/gh/prantlf/nettime/branch/master/graph/badge.svg)](https://codecov.io/gh/prantlf/nettime)
[![codebeat badge](https://codebeat.co/badges/9d85c898-df08-42fb-8ab9-407dc2ce2d22)](https://codebeat.co/projects/github-com-prantlf-/nettime-master)
![Dependency status](https://img.shields.io/librariesio/release/npm/nettime)

Prints time duration of various stages of a HTTP/S request, like DNS lookup, TLS handshake, Time to First Byte etc. Similarly to the [time] command, which measures process timings, the `nettime` command measures HTTP/S request timings.  You can find more information in [Understanding & Measuring HTTP Timings with Node.js](https://blog.risingstack.com/measuring-http-timings-node-js/).

**Attention**: Command-line options changed between 0.x and 1.x versions, so that they become compatible with [curl]. If you use the `nettime` command-line tool, check the affected options:

```text
-e, --ignore-certificate  =>  -k, --insecure
-u, --unit                =>  -t, --time-unit
-U, --user                =>  -u, --user
```

The programmatic interface did not change and has remained compatible.

## Command-line usage

Make sure that you have [Node.js] >= 14.8 installed. Install the `nettime` package globally and print timings of a sample web site:

```bash
$ npm install -g nettime
$ nettime https://www.google.com
Phase             Finished Duration
-----------------------------------
Socket Open         0.023s   0.023s
DNS Lookup          0.024s   0.001s
TCP Connection      0.053s   0.029s
TLS Handshake       0.133s   0.079s
First Byte          0.174s   0.041s
Content Transfer    0.176s   0.002s
Socket Close        0.177s   0.001s
-----------------------------------
Status Code: OK (200)
```

Running `nettime` without any parameters prints usage instructions:

```text
Usage: nettime [options] <URL>

Options:

  -V, --version                output the version number
  -0, --http1.0                use HTTP 1.0
      --http1.1                use HTTP 1.1 (default)
      --http2                  use HTTP 2.0
  -c, --connect-timeout <ms>   maximum time to wait for a connection
  -d, --data <data>            data to be sent using the POST verb
  -f, --format <format>        set output format: text, json, raw
  -H, --header <header>        send specific HTTP header
  -i, --include                include response headers in the output
  -I, --head                   use HEAD verb to get document info only
  -k, --insecure               ignore certificate errors
  -L, --location               follow redirects
  -o, --output <file>          write the received data to a file
  -t, --time-unit <unit>       set time unit: ms, s+ns
  -u, --user <credentials>     credentials for Basic Authentication
  -X, --request <verb>         specify HTTP verb to use for the request
  -C, --request-count <count>  count of requests to make (default: 1)
  -D, --request-delay <ms>     delay between two requests
  -A, --average-timings        print an average of multiple request timings
  -h, --help                   output usage information

The default output format is "text" and time unit "ms". Other options
are compatible with curl. Timings are printed to the standard output.

Examples:

  $ nettime -f json https://www.github.com
  $ nettime --http2 -C 3 -A https://www.google.com
```

## Programmatic usage

Make sure that you use [Node.js] >= 14.8. Install the `nettime` package locally and get time duration of waiting for the response and downloading the content of a sample web page:

```bash
npm install --save nettime
```

```js
const { nettime, getDuration } = require('nettime')
nettime('https://www.google.com')
  .then(result => {
    if (result.statusCode === 200) {
      let timings = result.timings
      let waiting = getDuration([0, 0], timings.firstByte)
      let downloading = getDuration(timings.firstByte, timings.contentTransfer)
      console.log('Waiting for the response:', nettime.getMilliseconds(waiting) + 'ms')
      console.log('Downloading the content:', nettime.getMilliseconds(downloading) + 'ms')
    }
  })
  .catch(error => console.error(error))
```

The main module exports a function which makes a HTTP/S request and returns a [Promise] to the result object. The function carries properties `nettime`, `getDuration`, `getMilliseconds` and `isRedirect`, so that the export can be consumed as an object with several static functions too:

```js
const nettime = require('nettime')
const { nettime, getDuration } = require('nettime')
```

The input argument is a string with a URL to make the request with, or an object with multiple properties.

The input object can contain:

* `url`: string with a URL to make the request with.
* `credentials`: object with `username` and `password` string properties to be used for formatting of the Basic Authentication HTTP header.
* `data`: string or Buffer to send to the server using the HTTP verb `POST` and the content type `application/x-www-form-urlencoded` by default.
* `failOnOutputFileError`: boolean for preventing the request timing operation from failing, if writing to the output file failed. If set to `false`, the error will be printed on the standard output and the process exit code will be set to 2. It is in effect only if `outputFile` is specified. The default is `true`.
* `headers`: object with header names as string keys and header values as string values.
* `httpVersion`: string with the protocol version ('1.0', '1.1' or '2.0') to be sent to the server. (Node.js HTTP support is hard-coded for 1.1. There can be a difference between 1.0 and 1.1 on the server side only. Node.js supports HTTP/2 in the version 8.4.0 or newer with the --expose-http2 command-lime option and in the version 8.8.1 or newer out-of-the-box. Alternatively, you can install a "http2" module as a polyfill.)
* `followRedirects`: boolean to continue making requests, if the original response contained a redirecting HTTP status code
* `includeHeaders`: boolean for including property `headers` (`Object`) with response headers in the promised result object. If `outputFile` is specified, the headers are written to the beginning of the output file too.
* `method`: HTTP verb to use in the HTTP request; `GET` is the default, unless `-i` or `-d` options are not set.
* `outputFile`: file path to write the received data to.
* `rejectUnauthorized`: boolean to refuse finishing the HTTPS request, is set to `true` (the default), if validation of the web site certificate fails; setting it to `false` makes the request ignore certificate errors.
* `returnResponse`: boolean for including property `response` (`Buffer`) with the received data in the promised result object.
* `requestCount`: integer for making multiple requests instead of one.
* `requestDelay`: integer to introduce a delay (in milliseconds ) between each two requests. The default is 100.
* `timeout`: intere to set the maximum time (in milliseconds) a single request should take before aborting it.

The result object contains:

* `httpVersion`: HTTP version, which the server responsed with (string).
* `statusCode`: [HTTP status code] of the response (integer).
* `statusMessage`: HTTP status message for the status code (string).
* `timings`: object with timing properties from various stages of the request. Timing is an array with two integers - seconds and nanoseconds passed since the request has been made, as returned by [process.hrtime].
* `headers`: an optional object with the response headers, if enabled by the option `includeHeaders`.
* `url`: an optional string with the requested URL, if the option `followRedirects` was set to `true`.

```js
{
  "httpVersion": '1.1',
  "statusCode": 200,
  "statusMessage": "OK",
  "timings": {
    "socketOpen": [ 0, 13260126 ],
    "dnsLookup": [ 0, 13747391 ],     // Optional, if hostname was specified
    "tcpConnection": [ 0, 152135165 ],
    "tlsHandshake": [ 0, 433219351 ], // Optional, if HTTPS protocol was used
    "firstByte": [ 1, 888887072 ],
    "contentTransfer": [ 1, 891221207 ],
    "socketClose": [ 1, 893156380 ]
  }
}
```

If the option `requestCount` is greater than `1`, the result objects will be returned in an array of the same length as the `requestCount` value.
If the option `followRedirects` us set to `true`, the result objects will be returned in an array of the length depending on the presence and count of redirecting responses.

*Note*: The `time-unit` parameter affects not only the "text" output format of the command line script, but also the "json" one. If set to "ms", timing values will be printed in milliseconds. If set to "s+ns", timings will be printed as arrays in [process.hrtime]'s format. Calling the `nettime` function programmatically will always return the timings as arrays in [process.hrtime]'s format.

### Helper functions

The following functions are exposed as named exports from the `nettime/lib/timings` module to help dealing with [process.hrtime]'s timing format and timings from multiple requests:

* `getDuration(start, end)`: computes the difference between two timings. Expects two arrays in [process.hrtime]'s format and returns the result as an array in the same [process.hrtime]'s format.
* `getMilliseconds(timing)`: converts the timing to milliseconds. Expects an array in [process.hrtime]'s format and returns the result as an integer.
* `computeAverageDurations(multipleTimings)`: computes average durations from an array of event timings. The array is supposed to contain objects with the same keys as the `timings` object from the `nettime` response. The returned object will contain the same keys pointing to event durations in [process.hrtime]'s format.
* `createTimingsFromDurations(timings, startTime)`: reconstructs event timings from event durations. The `timings` object is supposed to contain the same keys as the `timings` object from the `nettime` response, but pointing to event durations in [process.hrtime]'s format. The returned object will contain the same keys, but pointing to event times in [process.hrtime]'s format. The `startTime` parameter can shoft the event times. The default is no shift - `[0, 0]`.
* `isRedirect(statusCode)`: checks if the HTTP status code means a redirect. Returns `true` if it is, otherwise `false`.

These methods can be required separately too:

```js
const { isRedirect } = require('nettime')
const {
  getDuration, getMilliseconds,
  computeAverageDurations, createTimingsFromDurations
} = require('nettime/lib/timings')
```

Methods `getDuration`, `getMilliseconds` and `isRedirect` are accessible also as static methods of the `nettime` function exported from the main `nettime` module.

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style.  Add unit tests for any new or changed functionality. Lint and test your code using Grunt.

## License

Copyright (c) 2017-2022 Ferdinand Prantl

Licensed under the MIT license.

[time]: https://en.wikipedia.org/wiki/Time_(Unix)
[Node.js]: http://nodejs.org/
[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
[HTTP status code]: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes
[process.hrtime]: https://nodejs.org/api/process.html#process_process_hrtime_time
[curl]: https://curl.haxx.se/


CHANGELOG.md
# [5.0.0](https://github.com/prantlf/nettime/compare/v4.0.0...v5.0.0) (2022-12-20)


### Bug Fixes

* Replace commander with simple command-line parsing ([2f140c6](https://github.com/prantlf/nettime/commit/2f140c64fecb0d3289e76af669d8133af45b281a))


### Features

* Provide native ESM export ([6aa74c7](https://github.com/prantlf/nettime/commit/6aa74c754bdee75ee8566fc5a783330e85d77e7e))


### BREAKING CHANGES

* Although the same command-line syntax is supported, replacing the parser might cause a not integhded breaking change.
* Node.js 14.8 is required. Although this package should work with Node.js 12 or even Node.js 10 still well, the declaration of exports in package.json is recognised reliably by Node.js 14.8 or newer.

# [4.0.0](https://github.com/prantlf/nettime/compare/v3.0.1...v4.0.0) (2021-12-12)


### Features

* Add TypeScript types ([6f48b4a](https://github.com/prantlf/nettime/commit/6f48b4abfe6356f4413544748b3713ca4ede2e07))
* Allow following redirects ([0736037](https://github.com/prantlf/nettime/commit/0736037e0f79c76270a0a1b48786b53da5ecd060))


### chore

* Upgrade dependencies ([5dbbff7](https://github.com/prantlf/nettime/commit/5dbbff78cd82391775b23cd8b4751696c891313f))


### BREAKING CHANGES

*   Node.js has to be upgraded to 12 or newer version.



## [3.0.1](https://github.com/prantlf/nettime/compare/v3.0.0...v3.0.1) (2021-12-12)


### Bug Fixes

* Specifying timeout made the connection fail immediately ([cad761e](https://github.com/prantlf/nettime/commit/cad761e2f5181b0f199cc923ccbe7d48eeade279))



# [3.0.0](https://github.com/prantlf/nettime/compare/v2.1.4...v3.0.0) (2019-10-19)


### Features

* Add formatting the console output as a raw JSON (format "raw") ([f2f88f8](https://github.com/prantlf/nettime/commit/f2f88f8303ea87c1fbc97577127a4a0c7372cf71))
* Allow making multiple requests and returning their average timings ([112d581](https://github.com/prantlf/nettime/commit/112d581067fd2ebc5dab8c323be821831393ffb6))


### BREAKING CHANGES

* The source code was refactored to depend on some features available first in Node.js 8. Asynchronous async/await keywords, for example.



## [2.1.4](https://github.com/prantlf/nettime/compare/v2.1.3...v2.1.4) (2019-10-18)


### Bug Fixes

* Fix typo in a keyword ([671fb7b](https://github.com/prantlf/nettime/commit/671fb7bf24d3386138e13406c196fa3f9b3048f8))



## [2.1.3](https://github.com/prantlf/nettime/compare/v2.1.2...v2.1.3) (2019-10-18)


### Bug Fixes

* Upgrade package dependencies ([c608e0a](https://github.com/prantlf/nettime/commit/c608e0add1d7bf2f398f29324b58a787f3f735fe))
* Use a global  timeout handler a workaround for the idle socket setTimeout in Node.js 10+ ([1a4481f](https://github.com/prantlf/nettime/commit/1a4481ff15b41e243185bb7c0ed7f6b7cb91698f))



## [2.1.2](https://github.com/prantlf/nettime/compare/v2.1.1...v2.1.2) (2019-10-18)


### Bug Fixes

* Fix crash on Node.js 10+ caused by consuming both readable and data events ([7c86d3b](https://github.com/prantlf/nettime/commit/7c86d3bcead098aa4a660ab0219b904d60a780d0))
* Upgrade npm dependencies ([b3132b3](https://github.com/prantlf/nettime/commit/b3132b39fe0d4b4501a76031e0339c60a9dbaa4a))



## [2.1.1](https://github.com/prantlf/nettime/compare/v2.1.0...v2.1.1) (2019-06-08)


### Bug Fixes

* Upgrade module dependencies ([d54d700](https://github.com/prantlf/nettime/commit/d54d700c13c67a1bbe51dec8429aefd55c0525c4))



# [2.1.0](https://github.com/prantlf/nettime/compare/v2.0.1...v2.1.0) (2019-03-10)


### Bug Fixes

* Upgrade package dependencies ([415fe74](https://github.com/prantlf/nettime/commit/415fe74be7416e7b9549f27a60d8d13bcef27c43))


### Features

* Add support for specifying connection timeout ([6edb361](https://github.com/prantlf/nettime/commit/6edb361e1e8ebd992bc10569d5e4d71ee4676dce))



## [2.0.1](https://github.com/prantlf/nettime/compare/v2.0.0...v2.0.1) (2018-05-19)


### Bug Fixes

* Adapt http2 connection for Node.js 8.11.2 and Node.js 10 ([1663f64](https://github.com/prantlf/nettime/commit/1663f64fb189b951d968183aa506db31a776ed59))



# [2.0.0](https://github.com/prantlf/nettime/compare/v1.1.2...v2.0.0) (2018-04-27)


### Bug Fixes

* Upgrade NPM module dependencies ([8a464e0](https://github.com/prantlf/nettime/commit/8a464e01f59d0fc0441cec0119aa75b255ae776a))


### chore

* Dropped support of Node.js 4 ([5ac3a71](https://github.com/prantlf/nettime/commit/5ac3a7109344bbb8ffd599d1dbc8500bac3e1f07))


### BREAKING CHANGES

* Dropped support of Node.js 4



## [1.1.2](https://github.com/prantlf/nettime/compare/v1.1.1...v1.1.2) (2018-03-16)


### Bug Fixes

* Upgrade package dependencies ([02a440c](https://github.com/prantlf/nettime/commit/02a440c5dac16ffda5746c2434e58979467592fe))



## [1.1.1](https://github.com/prantlf/nettime/compare/v1.1.0...v1.1.1) (2017-12-21)


### Bug Fixes

* Upgrade semantic release and other dependencies ([5c0df8a](https://github.com/prantlf/nettime/commit/5c0df8ac334520bb5113029cd441f919a74752d6))



# [1.1.0](https://github.com/prantlf/nettime/compare/v1.0.0...v1.1.0) (2017-11-11)


### Bug Fixes

* Add code "ERR_INSECURE_SCHEME" to the error if https is not used for a HTTP/2 request ([5995015](https://github.com/prantlf/nettime/commit/5995015e6de72538abdb6078ed93ad0fa1a5cb48))
* Fail by default, if the operation if writing to the output file failed ([0dbf72c](https://github.com/prantlf/nettime/commit/0dbf72c1fd2c3940e61c1723e1b0db85f4841832))
* Return response headers independently on returning response content ([9af088f](https://github.com/prantlf/nettime/commit/9af088f4f652b58de16ca95611ecfb962a02a05f))


### Features

* Support (secure only) HTTP/2 requests ([a1f7a2f](https://github.com/prantlf/nettime/commit/a1f7a2f77b5b385032e0655f76693edcc1b603a7))
* Support specifying HTTP version 1.0 in the request header ([3cb303c](https://github.com/prantlf/nettime/commit/3cb303c9d3507e387a39abd47967bcdf99316010))



# [1.0.0](https://github.com/prantlf/nettime/compare/v0.5.0...v1.0.0) (2017-11-06)


### Features


tests/types.ts
import nettime, {
  nettime as nettime2, getDuration, getMilliseconds, isRedirect,
  NettimeResponse
} from '../lib/nettime.js'

const _r1: Promise<NettimeResponse> = nettime('')
const _r2: Promise<NettimeResponse> = nettime2({ url: '' })
const _d1: number = getDuration(1, 2)
const _m1: number = getMilliseconds([1])
const _i1: boolean = isRedirect(1)


lib/printer.js
import { events, getDuration } from './timings.js'
import { sprintf } from 'sprintf-js'

function printMilliseconds (timings) {
  let lastTiming = [0, 0]
  const output = [
    'Phase             Finished Duration',
    '-----------------------------------'
  ]
  for (const event in events) {
    const timing = timings[event]
    if (timing) {
      const timeFraction = Math.round(timing[1] / 1e6)
      const duration = getDuration(lastTiming, timing)
      const durationFraction = Math.round(duration[1] / 1e6)
      output.push(sprintf('%-17s %3d.%03ds %3d.%03ds',
        events[event], timing[0], timeFraction, duration[0], durationFraction))
      lastTiming = timing
    }
  }
  output.push('-----------------------------------')
  return output
}

function printNanoseconds (timings) {
  let lastTiming = [0, 0]
  const output = [
    'Phase             Finished       Duration',
    '-----------------------------------------------'
  ]
  for (const event in events) {
    const timing = timings[event]
    if (timing) {
      const timeFraction = Math.round(timing[1] / 1000) / 1000
      const duration = getDuration(lastTiming, timing)
      const durationFraction = Math.round(duration[1] / 1000) / 1000
      output.push(sprintf('%-17s %3ds %7.3fms %3ds %7.3fms',
        events[event], timing[0], timeFraction, duration[0], durationFraction))
      lastTiming = timing
    }
  }
  output.push('-----------------------------------------------')
  return output
}

export function printTimings (timings, timeUnit) {
  const print = timeUnit === 's+ns' ? printNanoseconds : printMilliseconds
  return print(timings).join('\n')
}


lib/timings.js
export const events = {
  socketOpen: 'Socket Open',
  dnsLookup: 'DNS Lookup',
  tcpConnection: 'TCP Connection',
  tlsHandshake: 'TLS Handshake',
  firstByte: 'First Byte',
  contentTransfer: 'Content Transfer',
  socketClose: 'Socket Close'
}

export function getDuration (start, end) {
  let seconds = end[0] - start[0]
  let nanoseconds = end[1] - start[1]
  if (nanoseconds < 0) {
    --seconds
    nanoseconds += 1e9
  }
  return [seconds, nanoseconds]
}

export function getMilliseconds ([seconds, nanoseconds]) {
  return seconds * 1000 + Math.round(nanoseconds / 1000) / 1000
}

export function computeAverageDurations (timings) {
  const timingCount = timings.length
  const durations = createEventDurations()
  computeEventDurations()
  checkSkippedEvents()
  computeEventDurationAverages()
  return durations

  function createEventDurations () {
    const durations = {}
    for (const event in events) {
      durations[event] = []
    }
    return durations
  }

  function computeEventDurations () {
    for (const timing of timings) {
      let lastTime = [0, 0]
      for (const event in events) {
        const time = timing[event]
        if (time) {
          const duration = getDuration(lastTime, time)
          durations[event].push(duration)
          lastTime = time
        } else {
          durations[event].push(undefined)
        }
      }
    }
  }

  function checkSkippedEvents () {
    for (const event in events) {
      const durationValues = durations[event]
      if (durationValues[0] === undefined) {
        for (let i = 0; i < timingCount; ++i) {
          if (durationValues[i] !== undefined) {
            throw new Error(`Unexpected event ${event} timing of the request ${i}.`)
          }
        }
      } else {
        for (let i = 0; i < timingCount; ++i) {
          if (durationValues[i] === undefined) {
            throw new Error(`Expected event ${event} timing of the request ${i}.`)
          }
        }
      }
    }
  }

  function computeEventDurationAverages () {
    for (const event in events) {
      durations[event] = durations[event].reduce((result, duration) =>
        duration ? [result[0] + duration[0], result[1] + duration[1]] : undefined,
      [0, 0])
    }
    for (const event in events) {
      const duration = durations[event]
      if (duration) {
        const [seconds, nanoseconds] = duration
        durations[event] = [seconds / timingCount, nanoseconds / timingCount]
      }
    }
  }
}

export function createTimingsFromDurations (durations, startTime) {
  const timings = {}
  let lastTiming = startTime || [0, 0]
  for (const event in events) {
    const duration = durations[event]
    if (duration) {
      const seconds = lastTiming[0] + duration[0]
      const nanoseconds = lastTiming[1] + duration[1]
      lastTiming = timings[event] = [seconds, nanoseconds]
    }
  }
  return timings
}


lib/nettime.d.ts
declare interface NettimeOptions {
  url: string
  credentials?: NettimeRequestCredentials
  method?: string
  headers?: Record<string, unknown>
  timeout?: number
  followRedirects?: boolean
  rejectUnauthorized?: boolean
  requestCount?: number
  requestDelay?: number
  outputFile?: string
  returnResponse?: boolean
  includeHeaders?: boolean
  appendToOutput?: unknown
  failOnOutputFileError?: boolean
  httpVersion?: string
  data?: unknown
}

declare interface NettimeRequestCredentials {
  username: string
  password: string
}

declare interface NettimeResponse {
  url?: string
  timings: NettimeTimings
  httpVersion: string
  statusCode: number
  statusMessage: string
  headers?: Record<string, unknown>
  response?: Buffer
}

declare interface NettimeTimings {
  socketOpen: number[]
  dnsLookup: number[]
  tcpConnection: number[]
  tlsHandshake: number[]
  firstByte: number[]
  contentTransfer: number[]
  socketClose: number[]
}

export default function nettime(options: string | NettimeOptions): Promise<NettimeResponse>
export function nettime(options: string | NettimeOptions): Promise<NettimeResponse>
export function getDuration (start: number, end: number): number
export function getMilliseconds (timings: number[]): number
export function isRedirect (statusCode: number): boolean


lib/nettime.js
import { writeFile } from 'fs'
import http from 'http'
import https from 'https'
import http2 from 'http2'
import { EOL } from 'os'
import { URL } from 'url'
import { getDuration, getMilliseconds } from './timings.js'

export async function nettime (options) {
  if (typeof options === 'string') {
    options = { url: options }
  }
  const { requestCount, requestDelay, followRedirects } = options
  const results = []
  if (requestCount > 1) {
    for (let i = 0; i < requestCount; ++i) {
      await makeRedirectableRequest()
      if (requestDelay) {
        await wait(requestDelay)
      }
      options.appendToOutput = true
    }
    return results
  }
  await makeRedirectableRequest()
  return followRedirects ? results : results[0]

  async function makeRedirectableRequest () {
    const { url: originalUrl } = options
    for (;;) {
      const { url } = options
      const result = await makeSingleRequest(options)
      if (followRedirects) {
        result.url = url
      }
      results.push(result)
      if (!(followRedirects && isRedirect(result.statusCode))) break
    }
    options.url = originalUrl
  }
}

function wait (milliseconds) {
  return new Promise(resolve => setTimeout(resolve, milliseconds))
}

function makeSingleRequest (options) {
  return new Promise((resolve, reject) => {
    const timings = {}
    const { outputFile, returnResponse, includeHeaders } = options
    let data = (outputFile || returnResponse) && Buffer.from([])
    let response

    function returnResult () {
      const result = { timings }
      if (response) {
        const { statusCode, headers } = response
        result.httpVersion = response.httpVersion
        result.statusCode = statusCode
        result.statusMessage = response.statusMessage
        if (includeHeaders) {
          result.headers = headers
        }
        // Prepare the next request
        if (isRedirect(statusCode)) {
          options.url = headers.location
        }
      }
      if (returnResponse && data) {
        result.response = data
      }
      resolve(result)
    }

    function writeOutputFile () {
      if (includeHeaders && response) {
        prependOutputHeader()
      }
      const flag = options.appendToOutput ? 'a' : 'w'
      return new Promise(resolve =>
        writeFile(outputFile, data, { flag }, error => {
          if (error) {
            if (options.failOnOutputFileError === false) {
              console.error(error.message)
              process.exitCode = 2
            } else {
              return reject(error)
            }
          }
          resolve()
        }))

      function prependOutputHeader () {
        const prolog = [`HTTP/${response.httpVersion} ${response.statusCode} ${response.statusMessage}`]
        const headers = response.headers
        if (headers) {
          const allHeaders = Object
            .keys(headers)
            .map(key => `${key}: ${headers[key]}`)
          Array.prototype.push.apply(prolog, allHeaders)
        }
        prolog.push(EOL)
        data = Buffer.concat([Buffer.from(prolog.join(EOL)), data])
      }
    }

    let firstByte
    function checkFirstByte () {
      if (!firstByte) {
        timings.firstByte = getTiming()
        firstByte = true
      }
    }

    let socketClosed
    function checkSocketClosed () {
      if (!socketClosed) {
        timings.socketClose = getTiming()
        if (outputFile && data) {
          writeOutputFile().then(returnResult)
        } else returnResult()
      }
    }

    function listenToSocket (socket) {
      timings.socketOpen = getTiming()
      socket
        .on('lookup', () => (timings.dnsLookup = getTiming()))
        .on('connect', () => (timings.tcpConnection = getTiming()))
        .on('secureConnect', () => (timings.tlsHandshake = getTiming()))
        .on('close', checkSocketClosed)
    }

    function listenToResponse (response) {
      response
        .on('readable', () => {
          checkFirstByte()
          const chunk = response.read()
          if (data && chunk !== null) {
            data = Buffer.concat([data, Buffer.from(chunk)])
          }
        })
        .on('end', () => (timings.contentTransfer = getTiming()))
    }

    const parameters = getParameters()
    const { httpVersion } = options
    const { protocol: scheme } = parameters
    let protocol
    if (httpVersion === '2.0') {
      if (scheme !== 'https:') {
        const error = new Error('HTTP/2 supports only the "https:" protocol.')
        error.code = 'ERR_INSECURE_SCHEME'
        throw error
      }
      protocol = http2
    } else {
      protocol = scheme === 'http:' ? http : https
    }

    const start = process.hrtime()

    const request = httpVersion === '2.0' ? makeHTTP2Request() : makeHTTP1Request()

    const timeout = options.timeout
    let timeoutHandler
    if (timeout) {
      request
        // Stopped working in Node.js 10. Added global setTimeout temporarily.
        /* c8 ignore next 16 */
        .on('timeout', () => {
          if (timeoutHandler) {
            clearTimeout(timeoutHandler)
          }
          request.abort()
          const error = new Error('Connection timed out.')
          error.code = 'ETIMEDOUT'
          reject(error)
        })
        .on('abort', () => {
          if (timeoutHandler) {
            const error = new Error('Connection timed out.')
            error.code = 'ETIMEDOUT'
            reject(error)
          }
        })
        .setTimeout(timeout)
    }

    const inputData = options.data
    if (inputData) {
      request.write(inputData)
    }
    request.end()

    // Workaround for Node.js 10+, which does not abort the connection attempt
    // any more, if it takes longer than the specified timeout. The local abort
    // handler will convert the abortion to a timeout in this case.
    if (timeout) {
      timeoutHandler = setTimeout(() => request.abort(), timeout)
    }

    function getParameters () {
      const { url, data, rejectUnauthorized, credentials } = options
      const headers = options.headers || {}
      const parameters = parseURL()
      setSecurity()
      setCredentials()
      setContentType()
      return parameters

      function parseURL () {
        const {
          protocol, username, password, host, hostname, port, pathname, search
        } = new URL(url)
        const auth = username
          ? password
              ? `${username}.${password}`
              : username
          : undefined
        const path = pathname ? pathname + search : undefined
        const method = options.method || (data ? 'POST' : 'GET')
        const agent = false
        return {
          protocol,
          username,
          password,
          auth,
          host,
          hostname,
          port,
          pathname,
          search,
          path,
          headers,
          method,
          agent
        }
      }

      function setSecurity () {
        if (rejectUnauthorized !== undefined) {
          parameters.rejectUnauthorized = rejectUnauthorized
        }
      }

      function setCredentials () {
        if (credentials) {
          const token = Buffer
            .from(`${credentials.username}:${credentials.password}`)
            .toString('base64')
          headers.authorization = `Basic ${token}`
        }
      }

      function setContentType () {
        if (data) {
          if (!headers['content-type']) {
            headers['content-type'] = 'application/x-www-form-urlencoded'
          }
          headers['content-length'] = Buffer.byteLength(data)
        }
      }
    }

    function makeHTTP2Request () {
      const origin = getOrigin()
      const rejectUnauthorized = parameters.rejectUnauthorized
      const client = protocol
        .connect(origin, { rejectUnauthorized })
        .on('socketError', reject)
        .on('error', reject)
      listenToSocket(client.socket)

      const headers = parameters.headers
      headers[':method'] = parameters.method
      headers[':path'] = parameters.pathname
      const request = client
        .request(headers)
        .on('response', headers => {
          const statusCode = headers[':status']
          const statusMessage = http.STATUS_CODES[statusCode]
          response = { headers, httpVersion, statusCode, statusMessage }
        })
      listenToResponse(request)
      request
        .on('end', () => client.close(checkSocketClosed))
        .setEncoding('utf8')
      return request
    }

    function makeHTTP1Request () {
      const request = protocol
        .request(parameters, localResponse => {
          listenToResponse(response = localResponse)
          response.setEncoding('utf8')
        })
        .on('socket', listenToSocket)
        .on('error', reject)
      if (httpVersion === '1.0') {
        enforceHTTP10()
      }
      return request

      function enforceHTTP10 () {
        const storeHeader = request._storeHeader
        request._storeHeader = (firstLine, headers) => {
          firstLine = firstLine.replace(/HTTP\/1.1\r\n$/, 'HTTP/1.0\r\n')
          return storeHeader.call(request, firstLine, headers)
        }
      }
    }

    function getOrigin () {
      const { hostname, port } = parameters
      let origin = `${scheme}//${hostname}`
      if (port) {
        origin += `:${port}`
      }
      return origin
    }

    function getTiming () {
      return getDuration(start, process.hrtime())
    }
  })
}

export function isRedirect (statusCode) {
  return statusCode >= 301 && statusCode <= 308 && statusCode !== 304
}

nettime.nettime = nettime
nettime.getDuration = getDuration
nettime.getMilliseconds = getMilliseconds
nettime.isRedirect = isRedirect

export { nettime as default, getDuration, getMilliseconds }
